pkg/cloudprovider/util/cloud_init_settings.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2021 The Machine Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"gopkg.in/yaml.v3"
+	"html/template"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	CloudInitNamespace = "cloud-init-settings"
+	jwtTokenNamePrefix = "cloud-init-getter-token"
+
+	bootstrapCloudConfig = `#cloud-config
```

Comment:
Unless you are planning to create a secret per node, I recommend that you set the node hostname using this cloud-init instead of the provisioning one.

Reply:
why? I won't to keep the script as clean as possible and also hostnames might differ for each cloud provider, which means adding conditions. The moment you start adding conditions, this means this is slim anymore.

Reply:
I agree with that in principle. However, in practice, it's just AWS that has a different behavior. The alternatives would be:
- Create a cloud-init-secret per node which is not really scalable.
- Write some magic that would be executed by the cloud-init-secret to figure out and apply the hostname. This will be tricky because you will need to figure out the hostname for kubelet configuration. It has to be "known".